File: Java/MajorityElement.java
/**
 * Boyer-Moore majority vote algorithm. Time complexity: O(n).
 */
public class MajorityElement {
    public static int majorityElement(int[] nums) {
        int candidate = Integer.MIN_VALUE;
        int count = 0;
        for (int num : nums) {
            if (count == 0) {
                candidate = num;
                count = 1;
            } else if (num == candidate) count++;
            else count--;
        }
        return candidate;
    }

    public static void main(String args[]) {
        System.out.println(majorityElement(new int[]{2, 3, 5, 3, 6, 3, 3, 4, 3})); // 3
    }
}

File: PythonCode/MajorityElement.py
from typing import List


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(1)
    # Boyer-Moore Voting Algorithm
    def majorityElement(self, nums: List[int]) -> int:
        majority = None
        count = 0
        for num in nums:
            if not majority or count == 0:
                majority = num
                count = 1
            elif majority == num:
                count += 1
            else:
                count -= 1
        return majority


if __name__ == '__main__':
    solution = Solution()
    print(solution.majorityElement([3, 2, 3]))              # 3
    print(solution.majorityElement([2, 2, 1, 1, 1, 2, 2]))  # 2
